sdks/java/io/couchbase/src/main/java/org/apache/beam/sdk/io/couchbase/CouchbaseIO.java

```diff
@@ -0,0 +1,359 @@
+    public Read withHttpPort(int port) {
+      checkArgument(port > 0, "httpPort must be > 0, but was: %s", port);
+      return builder().setHttpPort(port).build();
+    }
+
+    /**
+     * Define the carrier port connecting to Couchbase.
+     *
+     * @param port the carrier port
+     * @return a {@link PTransform} reading data from Couchbase
+     */
+    public Read withCarrierPort(int port) {
+      checkArgument(port > 0, "carrierPort must be > 0, but was: %s", port);
+      return builder().setCarrierPort(port).build();
+    }
+
+    /**
+     * Define the name of bucket.
+     *
+     * @param bucket the bucket name
+     * @return a {@link PTransform} reading data from Couchbase
+     */
+    public Read withBucket(String bucket) {
+      checkArgument(bucket != null, "bucket can not be null");
+      return builder().setBucket(bucket).build();
+    }
+
+    /**
+     * Define the bucket-level password to the target bucket.
+     *
+     * @param password password
+     * @return a {@link PTransform} reading data from Couchbase
+     */
+    public Read withPassword(String password) {
+      checkArgument(password != null, "password can not be null");
+      return builder().setPassword(password).build();
+    }
+
+    @Override
+    public PCollection<JsonDocument> expand(PBegin input) {
+      checkArgument((hosts() != null), "WithHosts()is required");
+      checkArgument(bucket() != null, "withBucket() is required");
+
+      CouchbaseSource source = new CouchbaseSource(this);
+      PCollection<JsonDocument> result = input.apply(org.apache.beam.sdk.io.Read.from(source));
+      // Disconnect the client from Couchbase
+      source.close();
+      return result;
+    }
+  }
+
+  @VisibleForTesting
+  static class CouchbaseSource extends BoundedSource<JsonDocument> {
+
+    private final Read spec;
+    private int itemCount;
+    private final int lowerBound; // Lower bound of key range (included)
+    private final int upperBound; // Upper bound of key range (excluded)
+    private Cluster cluster;
+    private Bucket bucket;
+
+    CouchbaseSource(Read spec) {
+      this(spec, null, null, 0, 0);
+    }
+
+    CouchbaseSource(Read spec, Cluster cluster, Bucket bucket, int lb, int ub) {
+      this.spec = spec;
+      this.cluster = cluster;
+      this.bucket = bucket;
+      this.lowerBound = lb;
+      this.upperBound = ub;
+    }
+
+    private Bucket getBucket() {
+      if (cluster == null) {
+        DefaultCouchbaseEnvironment.Builder builder = DefaultCouchbaseEnvironment.builder();
+        if (spec.httpPort() != null) {
+          builder.bootstrapHttpDirectPort(spec.httpPort());
+        }
+        if (spec.carrierPort() != null) {
+          builder.bootstrapCarrierDirectPort(spec.carrierPort());
+        }
+        cluster = CouchbaseCluster.create(builder.build(), spec.hosts());
+      }
+      if (bucket == null) {
+        // For Couchbase Server, in the previous version than 5.0, the passwordless bucket can be
```

Comment:
Please inline this comments, Beam's limit is 120 characters if I remember well so they should probably fit.